examples/arkode/C_serial/ark_lotka_volterra_ASA.c

```diff
@@ -0,0 +1,363 @@
+/* -----------------------------------------------------------------------------
+  * SUNDIALS Copyright Start
+  * Copyright (c) 2002-2024, Lawrence Livermore National Security
+  * and Southern Methodist University.
+  * All rights reserved.
+  *
+  * See the top-level LICENSE and NOTICE files for details.
+  *
+  * SPDX-License-Identifier: BSD-3-Clause
+  * SUNDIALS Copyright End
+  * -----------------------------------------------------------------------------
+  * This example solves the Lotka-Volterra ODE with four parameters,
+  *
+  *     u = [dx/dt] = [ p_0*x - p_1*x*y  ]
+  *         [dy/dt]   [ -p_2*y + p_3*x*y ].
+  *
+  * The initial condition is u(t_0) = 1.0 and we use the parameters
+  *  p  = [1.5, 1.0, 3.0, 1.0]. The integration interval can be controlled via
+  * the --tf command line argument, but by default it is t \in [0, 10.].
+  * An explicit Runge--Kutta method is employed via the ARKStep time stepper
+  * provided by ARKODE. After solving the forward problem, adjoint sensitivity
+  * analysis (ASA) is performed using the discrete adjoint method available with
+  * with ARKStep in order to obtain the gradient of the scalar cost function,
+  *
+  *    g(u(T),p) = sum_{i=1}^{n} u_{i}^2 / 2
+  *
+  * with respect to the initial condition and the parameters.
+  *
+  * ./ark_lotka_volterra_adj options:
+  * --tf <real>         the final simulation time
+  * --dt <real>         the timestep size
+  * --order <int>       the order of the RK method
+  * --check-freq <int>  how often to checkpoint (in steps)
+  * --no-stages         don't checkpoint stages
+  * --dont-keep         don't keep checkpoints around after loading
+  * --help              print these options
+  * ---------------------------------------------------------------------------*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sundials/sundials_core.h>
+
+#include <nvector/nvector_manyvector.h>
+#include <nvector/nvector_serial.h>
+#include <sunadjoint/sunadjoint_checkpointscheme_basic.h>
+#include <sunadjoint/sunadjoint_stepper.h>
+#include <sunmatrix/sunmatrix_dense.h>
+#include <sunmemory/sunmemory_system.h>
+
+#include <arkode/arkode.h>
+#include <arkode/arkode_arkstep.h>
+#include "sundials/sundials_nvector.h"
+
+typedef struct
+{
+  sunrealtype tf;
+  sunrealtype dt;
+  int order;
+  int check_freq;
+  sunbooleantype save_stages;
+  sunbooleantype keep_checks;
+} ProgramArgs;
+
+static sunrealtype params[4] = {1.5, 1.0, 3.0, 1.0};
```

Comment:
The standard specifies that the literals (constants) will be double precision. So, if using extended precision, the constants would be `double` and then casted to `long double` (to match `sunrealtype`).  In this case (for these values), I don't think that would be a problem. Since this is an example code, if it does not cause any issues, I think it is best to not use the `SUN_RCONST` macro (which isnt really intended for users) to avoid unnecessary clutter.

Reply:
`SUN_RCONST` is consistently used in other examples for literals. When we separate examples and tests we should revisit this but for now we should be consistent in its usage.